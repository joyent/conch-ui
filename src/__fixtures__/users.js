export const users = [
  {
    created: '2018-12-04T12:31:23.489677-08:00',
    email: 'user.name@joyent.com',
    force_password_change: true,
    id: '345a1234-9873-23ks-9983-221b8273629a',
    is_admin: false,
    last_login: '2019-01-23T15:59:59.041066-08:00',
    name: 'User Name',
    refuse_session_auth: true,
    workspaces: [
      {
        description: 'Master workspace',
        id: '434a98d1-5317-7h64-3b60-12fe533456m5',
        name: 'MASTER',
        parent_id: null,
        role: 'admin',
      },
      {
        description: 'DPC US-WEST-1 Expansion 1',
        id: 'as432423-e112-4bg8-6768-ab1ee54a6545',
        name: 'us-west-1-exp1',
        parent_id: '434a98d1-5317-7h64-3b60-12fe533456m5',
        role: 'admin',
        role_via: '434a98d1-5317-7h64-3b60-12fe533456m5',
      },
      {
        description: 'Conch Eng Dev',
        id: 'c16we45f-nd43-3yr9-94f1-232e42850odk',
        name: 'conch-eng-dev',
        parent_id: '434a98d1-5317-7h64-3b60-12fe533456m5',
        role: 'admin',
        role_via: '434a98d1-5317-7h64-3b60-12fe533456m5',
      },
      {
        description: 'Conch Dev Staging',
        id: 'b835a48d-928s-2j2f-9978-2323e23298a0',
        name: 'conch-dev-staging',
        parent_id: '434a98d1-5317-7h64-3b60-12fe533456m5',
        role: 'admin',
        role_via: '434a98d1-5317-7h64-3b60-12fe533456m5',
      },
    ],
  },
  {
    created: '2018-12-04T12:31:23.489677-08:00',
    email: 'another.user@joyent.com',
    force_password_change: false,
    id: '345a1234-9873-23ks-9983-221b8273629b',
    is_admin: true,
    last_login: null,
    name: 'Another User',
    refuse_session_auth: false,
    workspaces: [
      {
        description: 'Master workspace',
        id: '434a98d1-5317-7h64-3b60-12fe533456m5',
        name: 'MASTER',
        parent_id: null,
        role: 'admin',
      },
      {
        description: 'DPC US-WEST-1 Expansion 1',
        id: 'as432423-e112-4bg8-6768-ab1ee54a6545',
        name: 'us-west-1-exp1',
        parent_id: '434a98d1-5317-7h64-3b60-12fe533456m5',
        role: 'admin',
        role_via: '434a98d1-5317-7h64-3b60-12fe533456m5',
      },
      {
        description: 'Conch Eng Dev',
        id: 'c16we45f-nd43-3yr9-94f1-232e42850odk',
        name: 'conch-eng-dev',
        parent_id: '434a98d1-5317-7h64-3b60-12fe533456m5',
        role: 'admin',
        role_via: '434a98d1-5317-7h64-3b60-12fe533456m5',
      },
      {
        description: 'Conch Dev Staging',
        id: 'b835a48d-928s-2j2f-9978-2323e23298a0',
        name: 'conch-dev-staging',
        parent_id: '434a98d1-5317-7h64-3b60-12fe533456m5',
        role: 'admin',
        role_via: '434a98d1-5317-7h64-3b60-12fe533456m5',
      },
    ],
  },
];

export default users;
