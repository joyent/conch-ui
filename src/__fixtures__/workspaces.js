const workspaces = [
  {
    description: 'Master workspace',
    id: '434a98d1-5317-7h64-3b60-12fe533456m5',
    name: 'MASTER',
    parent_id: null,
    role: 'admin',
  },
  {
    description: 'DPC US-WEST-1 Expansion 1',
    id: 'as432423-e112-4bg8-6768-ab1ee54a6545',
    name: 'us-west-1-exp1',
    parent_id: '434a98d1-5317-7h64-3b60-12fe533456m5',
    role: 'admin',
    role_via: '434a98d1-5317-7h64-3b60-12fe533456m5',
  },
  {
    description: 'Conch Eng Dev',
    id: 'c16we45f-nd43-3yr9-94f1-232e42850odk',
    name: 'conch-eng-dev',
    parent_id: '434a98d1-5317-7h64-3b60-12fe533456m5',
    role: 'admin',
    role_via: '434a98d1-5317-7h64-3b60-12fe533456m5',
  },
  {
    description: 'Conch Dev Staging',
    id: 'b835a48d-928s-2j2f-9978-2323e23298a0',
    name: 'conch-dev-staging',
    parent_id: '434a98d1-5317-7h64-3b60-12fe533456m5',
    role: 'admin',
    role_via: '434a98d1-5317-7h64-3b60-12fe533456m5',
  },
  {
    description: 'Test Workspace',
    id: 'b835a48d-928s-2j2f-9978-2323e23298x9',
    name: 'test-workspace',
    parent_id: '434a98d1-5317-7h64-3b60-12fe533456m5',
    role: 'admin',
    role_via: '434a98d1-5317-7h64-3b60-12fe533456m5',
  },
];

export default workspaces;
