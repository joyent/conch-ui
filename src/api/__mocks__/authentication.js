export const isForcePasswordChange = () => {
  return Promise.resolve(true);
};

export const isLoggedIn = token =>
  token ? Promise.resolve(true) : Promise.reject(false);

export const login = data => {
  return new Promise((resolve, reject) => {
    const { user, password } = data;

    if (!user || !password) {
      reject({ error: '"user" and "password" are both required' });
    } else if (user !== 'validuser@joyent.com') {
      reject({ error: 'unauthorized' });
    } else if (password.length < 5) {
      reject({ error: 'Password must be at least 5 characters' });
    } else if (password !== 'goodPassword') {
      reject({ error: 'unauthorized' });
    }

    resolve({ jwt_token: true });
  });
};

export const logout = () => Promise.resolve(true);

export default {
  isForcePasswordChange,
  isLoggedIn,
  login,
  logout,
};
