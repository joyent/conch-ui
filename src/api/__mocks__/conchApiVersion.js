import { minimumApiVersion } from '@src/config.js';

export const getApiVersion = () => {
  return Promise.resolve({
    data: {
      version: minimumApiVersion,
    },
  });
};

export default { getApiVersion };
