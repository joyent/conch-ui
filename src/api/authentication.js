import {
  clearToken,
  getToken,
  setToken,
  request,
  requestWithToken,
} from './request.js';
import store from '@src/store/store.js';

export const isLoggedIn = () => !!getToken();

/**
 * Log in to Conch
 *
 * @param {String} email                    This is the user's email address
 * @param {String} password                 This is the user's password
 * @return {Response Object}
 */
export const login = (email, password) => {
  return request({
    method: 'POST',
    url: '/login',
    data: { email, password },
  }).then(response => {
    const data = response.data;

    if (data && data.jwt_token) {
      setToken(data.jwt_token);
    }

    store.dispatch('clearInvalidCredentials');

    return response;
  });
};

// TODO: As of this moment the server doesn't require authentication for `/logout` but it probably should
export const logout = () => {
  return requestWithToken({
    method: 'POST',
    url: '/logout',
  }).then(clearToken());
};

export default {
  isLoggedIn,
  login,
  logout,
};
