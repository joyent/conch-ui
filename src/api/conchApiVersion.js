import { request } from './request.js';

export const getApiVersion = () => {
  return request({
    method: 'GET',
    url: '/version',
  });
};

export default {
  getApiVersion,
};
