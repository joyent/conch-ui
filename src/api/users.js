import { clearToken, request, requestWithToken } from './request.js';

export const createUser = (user) => {
    return requestWithToken({
        method: 'POST',
        url: '/user',
        data: user,
    })
    .catch(error => {
        return Promise.reject(error);
    });
};

export const deactivateUser = (userId, clearTokens) => {
    return requestWithToken({
        method: 'DELETE',
        url: `/user/${userId}`,
        params: { clearTokens }
    });
};

export const demoteUser = (userId) => {
    return requestWithToken({
        method: 'POST',
        url: `/user/${userId}`,
        data: { is_admin: false },
    });
};

export const editUser = (user) => {
    const data = {};

    data.is_admin = user.is_admin;

    if (user.email) {
        data.email = user.email;
    }

    if (user.name) {
        data.name = user.name;
    }

    return requestWithToken({
        method: 'POST',
        url: `/user/${user.id}`,
        data
    })
    .catch(error => {
        return Promise.reject(error);
    });
};

export const forcePasswordChange = (userId) => {
    return requestWithToken({
        method: 'DELETE',
        url: `/user/${userId}/password`,
    });
};

export const getCurrentUser = () => {
    return request({
        method: 'GET',
        url: '/user/me',
    });
};

export const getUser = (userId) => {
    return request({
        method: 'GET',
        url: `/user/${userId}`,
    });
};

export const getUsers = () => {
    return requestWithToken({
        method: 'GET',
        url: '/user',
    });
};

export const promoteUser = (userId) => {
    return requestWithToken({
        method: 'POST',
        url: `/user/${userId}`,
        data: { is_admin: true },
    });
};

export const updatePassword = (password) => {
    return requestWithToken({
        method: 'POST',
        url: '/user/me/password',
        data: { password }
    })
    .then(clearToken());
};

// Tokens

export const createToken = (name) => {
    return requestWithToken({
        method: 'POST',
        url: '/user/me/token',
        data: { name },
    });
};

export const deleteToken = (name) => {
    return requestWithToken({
        method: 'DELETE',
        url: `/user/me/token/${name}`,
    });
};

// Used?
export const getToken = (name) => {
    return requestWithToken({
        method: 'GET',
        url: `/user/me/token/${name}`,
    });
};

// Profile page
export const getTokens = () => {
    return requestWithToken({
        method: 'GET',
        url: '/user/me/token',
    });
};

export const deleteUserToken = (name, userId) => {
    return requestWithToken({
        method: 'DELETE',
        url: `/user/${userId}/token/${name}`,
    });
};

export const deleteUserTokens = (userId, params) => {
    return requestWithToken({
        method: 'POST',
        url: `/user/${userId}/revoke`,
        params
    });
};

export const getUserTokens = (userId) => {
    return requestWithToken({
        method: 'GET',
        url: `/user/${userId}/token`,
    });
};


export default {
    createToken,
    deleteToken,
    getToken,
    getTokens,
    deleteUserToken,
    deleteUserTokens,

    createUser,
    deactivateUser,
    demoteUser,
    editUser,
    forcePasswordChange,
    getCurrentUser,
    getUser,
    getUsers,
    promoteUser,
    updatePassword,
};
