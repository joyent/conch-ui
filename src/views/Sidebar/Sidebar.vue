<template>
    <aside
        class="menu"
        :class="{ loading: !this.workspaceId || isEmpty(this.currentUser) }"
    >
        <Spinner v-if="!this.workspaceId || isEmpty(this.currentUser)" />
        <div class="sidebar" v-else>
            <div class="brand">
                <img
                    class="brand-icon"
                    src="../../assets/brand.png"
                    width="150px"
                    @click="navigateHomepage()"
                />
            </div>
            <p class="menu-label">Datacenter Builds</p>
            <ul class="menu-list">
                <li class="nav-item">
                    <router-link
                        :to="{
                            name: 'status',
                            params: { currentWorkspace: this.workspaceId },
                        }"
                        active-class="is-active"
                    >
                        <i class="material-icons">signal_cellular_alt</i>
                        <span>Status</span>
                    </router-link>
                </li>
                <li class="nav-item">
                    <router-link
                        :to="{
                            name: 'datacenter',
                            params: { currentWorkspace: this.workspaceId },
                        }"
                        active-class="is-active"
                    >
                        <i class="material-icons">search</i>
                        <span>Browse</span>
                    </router-link>
                </li>
                <li class="nav-item">
                    <router-link
                        :to="{
                            name: 'devices',
                            params: { currentWorkspace: this.workspaceId },
                        }"
                        active-class="is-active"
                    >
                        <i class="material-icons">dns</i>
                        <span>Devices</span>
                    </router-link>
                </li>
            </ul>
            <p class="menu-label">Pages</p>
            <ul
                class="sidenav-dropdown menu-list"
                :class="{ 'is-expanded': isBuildsExpanded }"
                @click="isBuildsExpanded = !isBuildsExpanded"
            >
                <li class="nav-item">
                    <a>
                        <i class="material-icons">layers</i>
                        <span class="name">Builds</span>
                        <span class="icon">
                            <i class="fas fa-chevron-right"></i>
                        </span>
                    </a>
                </li>
            </ul>
            <ul class="menu-list" v-if="isBuildsExpanded">
                <li class="nav-item">
                    <router-link
                        :to="{
                            name: 'build',
                            params: { buildId: 'a2dbe92ledsa99d' },
                        }"
                        active-class="is-active"
                    >
                        <span>UK-West-1</span>
                    </router-link>
                </li>
                <li class="nav-item">
                    <router-link
                        :to="{
                            name: 'build',
                            params: { buildId: 'b2wee92hhdsa99a' },
                        }"
                        active-class="is-active"
                    >
                        <span>US-NORTHWEST-1a</span>
                    </router-link>
                </li>
            </ul>
            <ul
                class="sidenav-dropdown menu-list"
                :class="{ 'is-expanded': isOrganizationsExpanded }"
                @click="isOrganizationsExpanded = !isOrganizationsExpanded"
            >
                <li class="nav-item">
                    <a>
                        <i class="material-icons">business</i>
                        <span class="name">Organizations</span>
                        <span class="icon">
                            <i class="fas fa-chevron-right"></i>
                        </span>
                    </a>
                </li>
            </ul>
            <ul class="menu-list" v-if="isOrganizationsExpanded">
                <li
                    class="nav-item"
                    v-for="organization in currentUser.organizations"
                    :key="organization.id"
                >
                    <router-link
                        :to="{
                            name: 'organization',
                            params: { organizationId: organization.id },
                        }"
                        active-class="is-active"
                        :key="`$route.path_${organization.id}`"
                    >
                        <span>{{ organization.name }}</span>
                    </router-link>
                </li>
            </ul>
            <p v-if="currentUser.is_admin" class="menu-label">Conch Admin</p>
            <ul class="menu-list" v-if="currentUser.is_admin">
                <li class="nav-item">
                    <router-link
                        :to="{ name: 'builds' }"
                        active-class="is-active"
                    >
                        <i class="material-icons">layers</i>
                        <span>Builds</span>
                    </router-link>
                </li>
                <li class="nav-item">
                    <router-link
                        :to="{ name: 'organizations' }"
                        active-class="is-active"
                    >
                        <i class="material-icons">business</i>
                        <span>Organizations</span>
                    </router-link>
                </li>
                <li class="nav-item">
                    <router-link
                        :to="{ name: 'tokens' }"
                        active-class="is-active"
                    >
                        <i class="material-icons">vpn_key</i>
                        <span>Tokens</span>
                    </router-link>
                </li>
                <li class="nav-item">
                    <router-link
                        :to="{ name: 'user-management' }"
                        active-class="is-active"
                    >
                        <i class="material-icons">people_alt</i>
                        <span>Users</span>
                    </router-link>
                </li>
            </ul>
            <p class="menu-label">Conch</p>
            <ul class="menu-list">
                <li class="nav-item">
                    <router-link
                        :to="{ name: 'user' }"
                        active-class="is-active"
                    >
                        <i class="material-icons">account_circle</i>
                        <span>Profile</span>
                    </router-link>
                </li>
                <li class="nav-item">
                    <a class="sign-out" @click="signOut()">
                        <i class="material-icons">exit_to_app</i>
                        <span>Sign Out</span>
                    </a>
                </li>
            </ul>
        </div>
    </aside>
</template>

<script>
import isEmpty from 'lodash/isEmpty';
import Spinner from '@src/views/components/Spinner.vue';
import { logout } from '@api/authentication.js';
import { getCurrentUser } from '@api/users.js';
import { mapActions, mapGetters, mapState } from 'vuex';

export default {
    components: {
        Spinner,
    },
    data() {
        return {
            conchVersion: '',
            conchUIVersion: '',
            isBuildsExpanded: false,
            isOrganizationsExpanded: false,
        };
    },
    methods: {
        ...mapActions(['setCurrentUser']),
        isEmpty,
        navigateHomepage() {
            this.$router.push({
                name: 'status',
                params: { currentWorkspace: this.currentWorkspaceId },
            });
        },
        signOut() {
            logout().then(() => {
                this.$router.push({ name: 'signIn' });
            });
        },
    },
    computed: {
        ...mapGetters(['currentWorkspaceId']),
        ...mapState(['currentUser', 'currentWorkspace']),
        workspaceId() {
            let workspaceId = this.currentWorkspaceId;

            if (!workspaceId && !isEmpty(this.currentWorkspace)) {
                workspaceId = this.currentWorkspace.id;
            }

            if (!workspaceId) {
                if (
                    this.$route &&
                    this.$route.params &&
                    this.$route.params.currentWorkspace
                ) {
                    workspaceId = this.$route.params.currentWorkspace;
                }
            }

            return workspaceId;
        },
    },
    created() {
        if (isEmpty(this.currentUser)) {
            getCurrentUser().then(response => {
                this.setCurrentUser(response.data);
            });
        }
    },
};
</script>
