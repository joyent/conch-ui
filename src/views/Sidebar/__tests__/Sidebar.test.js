import Sidebar from '../Sidebar.vue';
import Vuex from 'vuex';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import * as auth from '@api/authentication.js';
import * as conchApi from '@api/conchApiVersion.js';

// Fixture
import users from '@src/__fixtures__/users.js';
import workspaces from '@src/__fixtures__/workspaces.js';

const localVue = createLocalVue();
localVue.use(Vuex);

jest.mock('../../../api/request.js');
jest.spyOn(auth, 'logout');
jest.spyOn(conchApi, 'getApiVersion');

describe('Sidebar.vue', () => {
    let getters;
    let mocks;
    let state;
    let store;
    let stubs;
    let wrapper;

    beforeEach(() => {
        getters = { currentWorkspaceId: jest.fn() };
        mocks = { $router: [] };
        state = { currentUser: users[0], currentWorkspace: workspaces[0] };
        store = new Vuex.Store({ state, getters });
        stubs = ['router-link'];
        wrapper = shallowMount(Sidebar, { localVue, mocks, store, stubs });
    });

    test('should call signOut method when "Log out" link is clicked', () => {
        const spy = jest.spyOn(wrapper.vm, 'signOut');

        wrapper.find('.sign-out').trigger('click');

        return expect(spy).toHaveBeenCalled();
    });

    test('should call logout from auth API when signOut is called', () => {
        wrapper.find('.sign-out').trigger('click');

        return expect(auth.logout).toHaveBeenCalled();
    });

    test('should call getApiVersion from conchApiVersion API on created lifecycle hook', () => {
        return expect(conchApi.getApiVersion).toHaveBeenCalled();
    });

    test('should display a loading indicator when the currentWorkspace is being loaded', () => {
        state.currentWorkspace = {};
        store = new Vuex.Store({ state, getters });
        wrapper = shallowMount(Sidebar, { localVue, mocks, store, stubs });

        expect(wrapper.find('spinner-stub').exists()).toBeTruthy();
    });
});
