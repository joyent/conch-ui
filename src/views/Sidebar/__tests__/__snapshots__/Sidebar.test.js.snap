// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar.vue should render correctly on initial render 1`] = `
<aside class="menu">
  <div class="brand"><img src="../../assets/brand.png" width="150px" class="brand-icon"></div>
  <p class="menu-label">Builds</p>
  <ul class="menu-list">
    <li>
      <router-link-stub to="[object Object]" active-class="is-active"><i class="fas fa-lg fa-drafting-compass"></i>
        UK-West-1
      </router-link-stub>
    </li>
  </ul>
  <p class="menu-label">Datacenter Builds</p>
  <ul class="menu-list">
    <li class="nav-item">
      <router-link-stub to="[object Object]" active-class="is-active"><i class="fas fa-lg fa-satellite-dish"></i> <span>Status</span></router-link-stub>
    </li>
    <li class="nav-item">
      <router-link-stub to="[object Object]" active-class="is-active"><i class="fas fa-lg fa-search"></i> <span>Browse</span></router-link-stub>
    </li>
    <li class="nav-item">
      <router-link-stub to="[object Object]" active-class="is-active"><i class="fas fa-lg fa-server"></i> <span>Devices</span></router-link-stub>
    </li>
  </ul>
  <!---->
  <!---->
  <p class="menu-label">Conch</p>
  <ul class="menu-list">
    <li class="nav-item">
      <router-link-stub to="[object Object]" active-class="is-active"><i class="fas fa-lg fa-user"></i> <span>Profile</span></router-link-stub>
    </li>
    <li class="nav-item"><a class="sign-out"><i class="fas fa-lg fa-sign-out-alt"></i> <span>Log Out</span></a></li>
  </ul>
</aside>
`;
